jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -1537,115 +1554,105 @@ public void succeeded()
-                    !(totalWritten == 0 && (HttpMethod.HEAD.is(_request.getMethod()) || response.getStatus() == HttpStatus.NOT_MODIFIED_304)))
-                    failure = ExceptionUtil.combine(failure, new IOException("content-length %d != %d written".formatted(committedContentLength, totalWritten)));
-
-                // Is the request fully consumed?
-                Throwable unconsumed = stream.consumeAvailable();
-                if (LOG.isDebugEnabled())
-                    LOG.debug("consumeAvailable: {} {} ", unconsumed == null, httpChannelState);
-
-                if (unconsumed != null && httpChannelState.getConnectionMetaData().isPersistent())
-                    failure = ExceptionUtil.combine(failure, unconsumed);
-
-                if (failure != null)
+                // If we are still not failing, is a last stream send needed or can we complete?
+                if (failure == null)
                 {
+                    doLastStreamSend = httpChannelState.lockedLastStreamSend();
+                    if (doLastStreamSend)
+                        response._writeCallback = httpChannelState._handlerInvoker;
+                    // or complete the stream if everything is done.
+                    else if (httpChannelState.lockedIsLastStreamSendCompleted())
+                        completeStream = httpChannelState._handled;
+                }
+                else
+                {
+                    // We are failing...
                     httpChannelState._callbackFailure = failure;
-                    if (!stream.isCommitted())
+
+                    // Can we and should we generate an error response?
+                    if (!stream.isCommitted() && !ExceptionUtil.isAssociated(failure, Request.Handler.AbortException.class))
+                    {
+                        // We are not committed, so we can send an error response.
                         errorResponse = new ErrorResponse(request);
-                    else
-                        completeStream = true;
+                    }
+                    else if (httpChannelState._handled)
+                    {
+                        // Callback and handling are completed, so it is just a matter of the last write
+                        if (httpChannelState.lockedIsLastStreamSendCompleted())
+                        {
+                            // We are committed, handling completed, and last write completed, so complete the stream now.
+                            completeStream = true;
+                        }
+                        else if (response.lockedIsWriting())
+                        {
+                            // We are currently writing, so let the completion of that write handle the failure
+                            httpChannelState._callbackFailure = failure;
+                            failedCallback = response._writeCallback;
+                            response._writeCallback = httpChannelState._handlerInvoker;
```

Comment:
In this way, we may wait forever for the write to complete and invoke the `_handlerInvoker`.

How about this:


```suggestion
                            Runnable task = response.lockedFailWrite(failure);
                            failedCallback = Callback.from(task, httpChannelState._handlerInvoker);
```

In 12.1.x we will leverage the new `cancelSend()` feature automatically.

Reply:
I'm not sure that continuing on with a pending write pointing at the HttpChannelState is a good idea.  Feels like an invitation for a race.   What is wrong waiting for the write to complete.  It won't be forever unless they have disabled idle timeout.

Reply:
> I'm not sure that continuing on with a pending write pointing at the HttpChannelState is a good idea.

I'm not sure what you mean above.

My point is that failing the callback may return the buffer to the pool, while the write is still pending.
We fixed this in 12.1.x with `cancelSend()`, but we don't leverage it here, while we would if we call `lockedFailWrite()`.